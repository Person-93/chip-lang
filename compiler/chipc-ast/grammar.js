const INT_TYPES = [
  "u8",
  "u16",
  "u32",
  "u64",
  "u128",
  "usize",
  "i8",
  "i16",
  "i32",
  "i64",
  "i128",
  "isize",
];

const FLOAT_TYPES = ["f32", "f64"];

const NUMERIC_TYPES = [...INT_TYPES, ...FLOAT_TYPES];

module.exports = grammar({
  name: "chip",

  extras: $ => [/\s/], // TODO add line comments and block comments

  word: $ => $.word,

  rules: {
    file: $ => seq(repeat($._inner_attribute), repeat($._item)),

    _item: $ => choice(
      $.function,
      $.extern_block,
      // TODO other items
      //  - top-level attrs
      //  - static items (including constants)
      //  - imports
      //  - ADTs
      //  - impl blocks?
      //  - traits
    ),


    _inner_attribute: $ => seq("#^[", $.attr, "]"),
    _outer_attribute: $ => seq("#[", $.attr, "]"),
    attr: $ => seq(
      $.path,
      optional(choice(
        seq("=", field("value", $._attr_data_value)),
        seq(
          "(",
          field("args", sepBy(",", $._attr_data_value)),
          ")",
        ),
      )),
    ),
    _attr_data_value: $ => choice(
      $._literal,
      $.attr,
    ),


    function: $ => seq(
      repeat($._outer_attribute),
      optional($.visibility_modifier),
      optional($.function_modifier),
      "fn",
      // TODO generic params
      field("name", $.identifier),
      "(",
      field(
        "inputs",
        sepBy(",", seq(
          field("identifier", $.identifier),
          ":",
          field("type", $.type),
        )),
      ),
      ")",
      optional(seq("->", field("output", $.type))),
      ":",
      field("body", $._expr),
    ),
    function_modifier: $ => choice(
      "const",
      "unsafe",
      seq("extern", field("abi", optional($.string_literal))),
    ),


    type: $ => $.path,

    extern_block: $ => seq(
      "extern",
      field("abi", optional($.string_literal)),
      "{",
      repeat($._extern_item),
      "}",
    ),
    _extern_item: $ => seq(choice($.extern_function), ";"),
    extern_function: $ => seq(
      optional($.visibility_modifier),
      "fn",
      field("name", $.identifier),
      "(",
      field(
        "inputs",
        sepBy(",", seq(
          field("identifier", $.identifier),
          ":",
          field("type", $.type),
        )),
      ),
      ")",
      optional(seq("->", field("output", $.type))),
    ),


    visibility_modifier: $ => seq(
      "pub",
      optional(seq("(",
        field("restriction", choice("super", "package")),
        ")"),
      ),
    ),


    _expr: $ => choice(
      $._literal,
      $.tuple_expression,
      $.codeblock,
      // TODO unary operation
      // TODO binary operation
      // TODO member access
      // TODO cast operation
    ),
    codeblock: $ => seq(
      "{",
      repeat($.statement),
      field("trailing", optional($._expr)),
      "}",
    ),
    statement: $ => seq(choice($._expr, $.let_binding), ";"),

    tuple_expression: $ => seq(
      "(",
      field("members", sepBy1(",", $._expr)),
      ")",
    ),

    let_binding: $ => seq(
      "let",
      $.identifier,
      ":=",
      field("value", $._expr),
    ),


    path: $ => seq(optional("::"), sepBy1("::", $.path_segment, false)),
    path_segment: $ => seq($.identifier, optional($.generic_params)),
    identifier: $ => /[a-zA-Z_][a-zA-Z0-9_]*/,
    generic_params: $ => seq("::<", sepBy1(",", $._expr), ">"),


    _literal: $ => choice(
      $.unit_literal,
      $.string_literal,
      $.boolean_literal,
      $.numeric_literal,
    ),
    unit_literal: $ => token(seq("(", ")")),
    string_literal: $ => /"[^"]*"/, // TODO handle escape sequence in string literal
    boolean_literal: $ => choice("true", "false"),
    numeric_literal: $ => seq(
      field("value", $.digits),
      optional(seq(".", field("float_part", $.digits))),
      field("suffix", optional(choice(...NUMERIC_TYPES))),
    ),
    digits: $ => /\d+/,


    word: $ => /[_\p{XID_Start}][_\p{XID_Continue}]*/
  },
});

function sepBy(sep, rule, trailing = true) {
  return optional(sepBy1(sep, rule, trailing));
}

function sepBy1(sep, rule, trailing = true) {
  return trailing ?
    seq(rule, repeat(seq(sep, rule)), optional(sep)) :
    seq(rule, repeat(seq(sep, rule)));
}
