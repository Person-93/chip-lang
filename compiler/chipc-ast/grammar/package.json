{
  "dependencies": {
    "nan": "^2.12.1"
  },
  "description": "chip grammar for tree-sitter",
  "devDependencies": {
    "tree-sitter-cli": "^0.20.6"
  },
  "license": "UNLICENSED",
  "main": "bindings/node",
  "name": "tree-sitter-chip",
  "scripts": {
    "generate": "tree-sitter generate --no-bindings ../grammar.js"
  },
  "version": "0.0.0"
}
