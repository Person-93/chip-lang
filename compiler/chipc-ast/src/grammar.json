{
  "name": "chip",
  "word": "word",
  "rules": {
    "file": {
      "type": "SEQ",
      "members": [
        {
          "type": "REPEAT",
          "content": {
            "type": "SYMBOL",
            "name": "inner_attribute"
          }
        },
        {
          "type": "REPEAT",
          "content": {
            "type": "SYMBOL",
            "name": "item"
          }
        }
      ]
    },
    "item": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SYMBOL",
          "name": "function"
        },
        {
          "type": "SYMBOL",
          "name": "extern_block"
        }
      ]
    },
    "inner_attribute": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "#^["
        },
        {
          "type": "SYMBOL",
          "name": "attr"
        },
        {
          "type": "STRING",
          "value": "]"
        }
      ]
    },
    "outer_attribute": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "#["
        },
        {
          "type": "SYMBOL",
          "name": "attr"
        },
        {
          "type": "STRING",
          "value": "]"
        }
      ]
    },
    "attr": {
      "type": "SEQ",
      "members": [
        {
          "type": "SYMBOL",
          "name": "path"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "SEQ",
                  "members": [
                    {
                      "type": "STRING",
                      "value": "="
                    },
                    {
                      "type": "FIELD",
                      "name": "value",
                      "content": {
                        "type": "SYMBOL",
                        "name": "attr_data_value"
                      }
                    }
                  ]
                },
                {
                  "type": "SEQ",
                  "members": [
                    {
                      "type": "STRING",
                      "value": "("
                    },
                    {
                      "type": "CHOICE",
                      "members": [
                        {
                          "type": "SEQ",
                          "members": [
                            {
                              "type": "FIELD",
                              "name": "arg",
                              "content": {
                                "type": "SYMBOL",
                                "name": "attr_data_value"
                              }
                            },
                            {
                              "type": "REPEAT",
                              "content": {
                                "type": "SEQ",
                                "members": [
                                  {
                                    "type": "STRING",
                                    "value": ","
                                  },
                                  {
                                    "type": "FIELD",
                                    "name": "arg",
                                    "content": {
                                      "type": "SYMBOL",
                                      "name": "attr_data_value"
                                    }
                                  }
                                ]
                              }
                            },
                            {
                              "type": "CHOICE",
                              "members": [
                                {
                                  "type": "STRING",
                                  "value": ","
                                },
                                {
                                  "type": "BLANK"
                                }
                              ]
                            }
                          ]
                        },
                        {
                          "type": "BLANK"
                        }
                      ]
                    },
                    {
                      "type": "STRING",
                      "value": ")"
                    }
                  ]
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        }
      ]
    },
    "attr_data_value": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SYMBOL",
          "name": "literal"
        },
        {
          "type": "SYMBOL",
          "name": "attr"
        }
      ]
    },
    "function": {
      "type": "SEQ",
      "members": [
        {
          "type": "REPEAT",
          "content": {
            "type": "SYMBOL",
            "name": "outer_attribute"
          }
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "FIELD",
              "name": "vis",
              "content": {
                "type": "SYMBOL",
                "name": "visibility_modifier"
              }
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "FIELD",
              "name": "modifier",
              "content": {
                "type": "SYMBOL",
                "name": "function_modifier"
              }
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": "fn"
        },
        {
          "type": "FIELD",
          "name": "name",
          "content": {
            "type": "SYMBOL",
            "name": "identifier"
          }
        },
        {
          "type": "SYMBOL",
          "name": "inputs"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SYMBOL",
              "name": "output"
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": ":"
        },
        {
          "type": "FIELD",
          "name": "body",
          "content": {
            "type": "SYMBOL",
            "name": "expr"
          }
        }
      ]
    },
    "function_modifier": {
      "type": "CHOICE",
      "members": [
        {
          "type": "STRING",
          "value": "const"
        },
        {
          "type": "STRING",
          "value": "unsafe"
        },
        {
          "type": "ALIAS",
          "content": {
            "type": "SEQ",
            "members": [
              {
                "type": "STRING",
                "value": "extern"
              },
              {
                "type": "CHOICE",
                "members": [
                  {
                    "type": "FIELD",
                    "name": "abi",
                    "content": {
                      "type": "SYMBOL",
                      "name": "string_literal"
                    }
                  },
                  {
                    "type": "BLANK"
                  }
                ]
              }
            ]
          },
          "named": true,
          "value": "extern"
        }
      ]
    },
    "extern_block": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "extern"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "FIELD",
              "name": "abi",
              "content": {
                "type": "SYMBOL",
                "name": "string_literal"
              }
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "STRING",
              "value": "{"
            },
            {
              "type": "REPEAT",
              "content": {
                "type": "FIELD",
                "name": "item",
                "content": {
                  "type": "SYMBOL",
                  "name": "extern_item"
                }
              }
            },
            {
              "type": "STRING",
              "value": "}"
            }
          ]
        }
      ]
    },
    "extern_item": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SYMBOL",
          "name": "extern_function"
        }
      ]
    },
    "extern_function": {
      "type": "SEQ",
      "members": [
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "FIELD",
              "name": "vis",
              "content": {
                "type": "SYMBOL",
                "name": "visibility_modifier"
              }
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": "fn"
        },
        {
          "type": "FIELD",
          "name": "name",
          "content": {
            "type": "SYMBOL",
            "name": "identifier"
          }
        },
        {
          "type": "SYMBOL",
          "name": "inputs"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SYMBOL",
              "name": "output"
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": ";"
        }
      ]
    },
    "inputs": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "("
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "FIELD",
                  "name": "input",
                  "content": {
                    "type": "SYMBOL",
                    "name": "annotated"
                  }
                },
                {
                  "type": "REPEAT",
                  "content": {
                    "type": "SEQ",
                    "members": [
                      {
                        "type": "STRING",
                        "value": ","
                      },
                      {
                        "type": "FIELD",
                        "name": "input",
                        "content": {
                          "type": "SYMBOL",
                          "name": "annotated"
                        }
                      }
                    ]
                  }
                },
                {
                  "type": "CHOICE",
                  "members": [
                    {
                      "type": "STRING",
                      "value": ","
                    },
                    {
                      "type": "BLANK"
                    }
                  ]
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": ")"
        }
      ]
    },
    "annotated": {
      "type": "SEQ",
      "members": [
        {
          "type": "SYMBOL",
          "name": "identifier"
        },
        {
          "type": "STRING",
          "value": ":"
        },
        {
          "type": "SYMBOL",
          "name": "type"
        }
      ]
    },
    "output": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "->"
        },
        {
          "type": "FIELD",
          "name": "output",
          "content": {
            "type": "SYMBOL",
            "name": "type"
          }
        }
      ]
    },
    "visibility_modifier": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "pub"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "STRING",
                  "value": "("
                },
                {
                  "type": "FIELD",
                  "name": "restriction",
                  "content": {
                    "type": "CHOICE",
                    "members": [
                      {
                        "type": "STRING",
                        "value": "super"
                      },
                      {
                        "type": "STRING",
                        "value": "package"
                      }
                    ]
                  }
                },
                {
                  "type": "STRING",
                  "value": ")"
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        }
      ]
    },
    "type": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SYMBOL",
          "name": "basic_type"
        },
        {
          "type": "SYMBOL",
          "name": "tuple_type"
        },
        {
          "type": "SYMBOL",
          "name": "function_type"
        },
        {
          "type": "STRING",
          "value": "Self"
        }
      ]
    },
    "basic_type": {
      "type": "SEQ",
      "members": [
        {
          "type": "SYMBOL",
          "name": "path"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SYMBOL",
              "name": "generic_args"
            },
            {
              "type": "BLANK"
            }
          ]
        }
      ]
    },
    "unit_type": {
      "type": "TOKEN",
      "content": {
        "type": "SEQ",
        "members": [
          {
            "type": "STRING",
            "value": "("
          },
          {
            "type": "STRING",
            "value": ")"
          }
        ]
      }
    },
    "tuple_type": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "("
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "SYMBOL",
              "name": "type"
            },
            {
              "type": "REPEAT",
              "content": {
                "type": "SEQ",
                "members": [
                  {
                    "type": "STRING",
                    "value": ","
                  },
                  {
                    "type": "SYMBOL",
                    "name": "type"
                  }
                ]
              }
            },
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "STRING",
                  "value": ","
                },
                {
                  "type": "BLANK"
                }
              ]
            }
          ]
        },
        {
          "type": "STRING",
          "value": ")"
        }
      ]
    },
    "function_type": {
      "type": "SEQ",
      "members": [
        {
          "type": "SYMBOL",
          "name": "function_modifier"
        },
        {
          "type": "STRING",
          "value": "fn"
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "STRING",
              "value": "("
            },
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "SEQ",
                  "members": [
                    {
                      "type": "FIELD",
                      "name": "input",
                      "content": {
                        "type": "SYMBOL",
                        "name": "type"
                      }
                    },
                    {
                      "type": "REPEAT",
                      "content": {
                        "type": "SEQ",
                        "members": [
                          {
                            "type": "STRING",
                            "value": ","
                          },
                          {
                            "type": "FIELD",
                            "name": "input",
                            "content": {
                              "type": "SYMBOL",
                              "name": "type"
                            }
                          }
                        ]
                      }
                    },
                    {
                      "type": "CHOICE",
                      "members": [
                        {
                          "type": "STRING",
                          "value": ","
                        },
                        {
                          "type": "BLANK"
                        }
                      ]
                    }
                  ]
                },
                {
                  "type": "BLANK"
                }
              ]
            },
            {
              "type": "STRING",
              "value": ")"
            }
          ]
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "STRING",
                  "value": "->"
                },
                {
                  "type": "FIELD",
                  "name": "output",
                  "content": {
                    "type": "SYMBOL",
                    "name": "type"
                  }
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        }
      ]
    },
    "type_projection": {
      "type": "SEQ",
      "members": [
        {
          "type": "FIELD",
          "name": "base",
          "content": {
            "type": "SYMBOL",
            "name": "type"
          }
        },
        {
          "type": "STRING",
          "value": "::"
        },
        {
          "type": "SYMBOL",
          "name": "identifier"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SYMBOL",
              "name": "generic_args"
            },
            {
              "type": "BLANK"
            }
          ]
        }
      ]
    },
    "generic_params": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "<"
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "SYMBOL",
              "name": "generic_param"
            },
            {
              "type": "REPEAT",
              "content": {
                "type": "SEQ",
                "members": [
                  {
                    "type": "STRING",
                    "value": ","
                  },
                  {
                    "type": "SYMBOL",
                    "name": "generic_param"
                  }
                ]
              }
            },
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "STRING",
                  "value": ","
                },
                {
                  "type": "BLANK"
                }
              ]
            }
          ]
        },
        {
          "type": "STRING",
          "value": ">"
        }
      ]
    },
    "generic_param": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SEQ",
          "members": [
            {
              "type": "STRING",
              "value": "const"
            },
            {
              "type": "FIELD",
              "name": "name",
              "content": {
                "type": "SYMBOL",
                "name": "identifier"
              }
            },
            {
              "type": "STRING",
              "value": ":"
            },
            {
              "type": "SYMBOL",
              "name": "type"
            },
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "SEQ",
                  "members": [
                    {
                      "type": "STRING",
                      "value": "="
                    },
                    {
                      "type": "FIELD",
                      "name": "default",
                      "content": {
                        "type": "SYMBOL",
                        "name": "expr"
                      }
                    }
                  ]
                },
                {
                  "type": "BLANK"
                }
              ]
            }
          ]
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "FIELD",
              "name": "name",
              "content": {
                "type": "SYMBOL",
                "name": "identifier"
              }
            },
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "SEQ",
                  "members": [
                    {
                      "type": "STRING",
                      "value": "="
                    },
                    {
                      "type": "FIELD",
                      "name": "default",
                      "content": {
                        "type": "SYMBOL",
                        "name": "type"
                      }
                    }
                  ]
                },
                {
                  "type": "BLANK"
                }
              ]
            }
          ]
        }
      ]
    },
    "generic_args": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "::<"
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "SYMBOL",
                  "name": "infer"
                },
                {
                  "type": "FIELD",
                  "name": "value",
                  "content": {
                    "type": "SYMBOL",
                    "name": "expr"
                  }
                }
              ]
            },
            {
              "type": "REPEAT",
              "content": {
                "type": "SEQ",
                "members": [
                  {
                    "type": "STRING",
                    "value": ","
                  },
                  {
                    "type": "CHOICE",
                    "members": [
                      {
                        "type": "SYMBOL",
                        "name": "infer"
                      },
                      {
                        "type": "FIELD",
                        "name": "value",
                        "content": {
                          "type": "SYMBOL",
                          "name": "expr"
                        }
                      }
                    ]
                  }
                ]
              }
            },
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "STRING",
                  "value": ","
                },
                {
                  "type": "BLANK"
                }
              ]
            }
          ]
        },
        {
          "type": "STRING",
          "value": ">"
        }
      ]
    },
    "path": {
      "type": "SEQ",
      "members": [
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "CHOICE",
                  "members": [
                    {
                      "type": "STRING",
                      "value": "::"
                    },
                    {
                      "type": "BLANK"
                    }
                  ]
                },
                {
                  "type": "FIELD",
                  "name": "segment",
                  "content": {
                    "type": "SYMBOL",
                    "name": "identifier"
                  }
                }
              ]
            },
            {
              "type": "STRING",
              "value": "package"
            },
            {
              "type": "STRING",
              "value": "self"
            },
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "STRING",
                  "value": "super"
                },
                {
                  "type": "REPEAT",
                  "content": {
                    "type": "SEQ",
                    "members": [
                      {
                        "type": "STRING",
                        "value": "::"
                      },
                      {
                        "type": "STRING",
                        "value": "super"
                      }
                    ]
                  }
                }
              ]
            }
          ]
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "STRING",
                  "value": "::"
                },
                {
                  "type": "SEQ",
                  "members": [
                    {
                      "type": "FIELD",
                      "name": "segment",
                      "content": {
                        "type": "SYMBOL",
                        "name": "identifier"
                      }
                    },
                    {
                      "type": "REPEAT",
                      "content": {
                        "type": "SEQ",
                        "members": [
                          {
                            "type": "STRING",
                            "value": "::"
                          },
                          {
                            "type": "FIELD",
                            "name": "segment",
                            "content": {
                              "type": "SYMBOL",
                              "name": "identifier"
                            }
                          }
                        ]
                      }
                    }
                  ]
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        }
      ]
    },
    "expr": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SYMBOL",
          "name": "literal"
        },
        {
          "type": "SYMBOL",
          "name": "tuple_expression"
        },
        {
          "type": "SYMBOL",
          "name": "codeblock"
        }
      ]
    },
    "codeblock": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "{"
        },
        {
          "type": "SEQ",
          "members": [
            {
              "type": "REPEAT",
              "content": {
                "type": "SYMBOL",
                "name": "statement"
              }
            },
            {
              "type": "FIELD",
              "name": "trailing",
              "content": {
                "type": "CHOICE",
                "members": [
                  {
                    "type": "SYMBOL",
                    "name": "expr"
                  },
                  {
                    "type": "BLANK"
                  }
                ]
              }
            }
          ]
        },
        {
          "type": "STRING",
          "value": "}"
        }
      ]
    },
    "statement": {
      "type": "SEQ",
      "members": [
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "CHOICE",
              "members": [
                {
                  "type": "SYMBOL",
                  "name": "let_binding"
                },
                {
                  "type": "SYMBOL",
                  "name": "expr"
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": ";"
        }
      ]
    },
    "tuple_expression": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "("
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "FIELD",
                  "name": "member",
                  "content": {
                    "type": "SYMBOL",
                    "name": "expr"
                  }
                },
                {
                  "type": "REPEAT",
                  "content": {
                    "type": "SEQ",
                    "members": [
                      {
                        "type": "STRING",
                        "value": ","
                      },
                      {
                        "type": "FIELD",
                        "name": "member",
                        "content": {
                          "type": "SYMBOL",
                          "name": "expr"
                        }
                      }
                    ]
                  }
                },
                {
                  "type": "CHOICE",
                  "members": [
                    {
                      "type": "STRING",
                      "value": ","
                    },
                    {
                      "type": "BLANK"
                    }
                  ]
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": ")"
        }
      ]
    },
    "let_binding": {
      "type": "SEQ",
      "members": [
        {
          "type": "STRING",
          "value": "let"
        },
        {
          "type": "SYMBOL",
          "name": "identifier"
        },
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "STRING",
                  "value": ":"
                },
                {
                  "type": "SYMBOL",
                  "name": "type"
                }
              ]
            },
            {
              "type": "BLANK"
            }
          ]
        },
        {
          "type": "STRING",
          "value": ":="
        },
        {
          "type": "FIELD",
          "name": "value",
          "content": {
            "type": "SYMBOL",
            "name": "expr"
          }
        }
      ]
    },
    "identifier": {
      "type": "PATTERN",
      "value": "[a-zA-Z_][a-zA-Z0-9_]*"
    },
    "infer": {
      "type": "STRING",
      "value": "_"
    },
    "literal": {
      "type": "CHOICE",
      "members": [
        {
          "type": "SYMBOL",
          "name": "string_literal"
        },
        {
          "type": "SYMBOL",
          "name": "boolean_literal"
        },
        {
          "type": "SYMBOL",
          "name": "numeric_literal"
        }
      ]
    },
    "string_literal": {
      "type": "PATTERN",
      "value": "\"[^\"]*\""
    },
    "boolean_literal": {
      "type": "CHOICE",
      "members": [
        {
          "type": "STRING",
          "value": "true"
        },
        {
          "type": "STRING",
          "value": "false"
        }
      ]
    },
    "numeric_literal": {
      "type": "SEQ",
      "members": [
        {
          "type": "CHOICE",
          "members": [
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "FIELD",
                  "name": "digits",
                  "content": {
                    "type": "SYMBOL",
                    "name": "digits"
                  }
                },
                {
                  "type": "CHOICE",
                  "members": [
                    {
                      "type": "SEQ",
                      "members": [
                        {
                          "type": "STRING",
                          "value": "."
                        },
                        {
                          "type": "FIELD",
                          "name": "float_part",
                          "content": {
                            "type": "SYMBOL",
                            "name": "digits"
                          }
                        }
                      ]
                    },
                    {
                      "type": "BLANK"
                    }
                  ]
                }
              ]
            },
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "FIELD",
                  "name": "prefix",
                  "content": {
                    "type": "STRING",
                    "value": "0x"
                  }
                },
                {
                  "type": "FIELD",
                  "name": "digits",
                  "content": {
                    "type": "SYMBOL",
                    "name": "hex_digits"
                  }
                },
                {
                  "type": "CHOICE",
                  "members": [
                    {
                      "type": "SEQ",
                      "members": [
                        {
                          "type": "STRING",
                          "value": "."
                        },
                        {
                          "type": "FIELD",
                          "name": "float_part",
                          "content": {
                            "type": "SYMBOL",
                            "name": "hex_digits"
                          }
                        }
                      ]
                    },
                    {
                      "type": "BLANK"
                    }
                  ]
                }
              ]
            },
            {
              "type": "SEQ",
              "members": [
                {
                  "type": "FIELD",
                  "name": "prefix",
                  "content": {
                    "type": "STRING",
                    "value": "0b"
                  }
                },
                {
                  "type": "FIELD",
                  "name": "digits",
                  "content": {
                    "type": "SYMBOL",
                    "name": "bin_digits"
                  }
                },
                {
                  "type": "CHOICE",
                  "members": [
                    {
                      "type": "SEQ",
                      "members": [
                        {
                          "type": "STRING",
                          "value": "."
                        },
                        {
                          "type": "FIELD",
                          "name": "float_part",
                          "content": {
                            "type": "SYMBOL",
                            "name": "bin_digits"
                          }
                        }
                      ]
                    },
                    {
                      "type": "BLANK"
                    }
                  ]
                }
              ]
            }
          ]
        },
        {
          "type": "FIELD",
          "name": "suffix",
          "content": {
            "type": "CHOICE",
            "members": [
              {
                "type": "CHOICE",
                "members": [
                  {
                    "type": "STRING",
                    "value": "u8"
                  },
                  {
                    "type": "STRING",
                    "value": "u16"
                  },
                  {
                    "type": "STRING",
                    "value": "u32"
                  },
                  {
                    "type": "STRING",
                    "value": "u64"
                  },
                  {
                    "type": "STRING",
                    "value": "u128"
                  },
                  {
                    "type": "STRING",
                    "value": "usize"
                  },
                  {
                    "type": "STRING",
                    "value": "i8"
                  },
                  {
                    "type": "STRING",
                    "value": "i16"
                  },
                  {
                    "type": "STRING",
                    "value": "i32"
                  },
                  {
                    "type": "STRING",
                    "value": "i64"
                  },
                  {
                    "type": "STRING",
                    "value": "i128"
                  },
                  {
                    "type": "STRING",
                    "value": "isize"
                  },
                  {
                    "type": "STRING",
                    "value": "f32"
                  },
                  {
                    "type": "STRING",
                    "value": "f64"
                  }
                ]
              },
              {
                "type": "BLANK"
              }
            ]
          }
        }
      ]
    },
    "digits": {
      "type": "PATTERN",
      "value": "\\d+"
    },
    "bin_digits": {
      "type": "PATTERN",
      "value": "[01]+"
    },
    "hex_digits": {
      "type": "PATTERN",
      "value": "[\\da-f]+"
    },
    "word": {
      "type": "PATTERN",
      "value": "[_\\p{XID_Start}][_\\p{XID_Continue}]*"
    }
  },
  "extras": [
    {
      "type": "PATTERN",
      "value": "\\s"
    }
  ],
  "conflicts": [],
  "precedences": [],
  "externals": [],
  "inline": [
    "attr_data_value",
    "expr",
    "extern_item",
    "generic_args",
    "generic_params",
    "inner_attribute",
    "inputs",
    "item",
    "literal",
    "outer_attribute",
    "output",
    "type"
  ],
  "supertypes": []
}
