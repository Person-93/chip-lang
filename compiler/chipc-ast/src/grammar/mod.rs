use tree_sitter::Language;

#[cfg(test)]
mod tests;

pub fn language() -> Language {
  unsafe { tree_sitter_chip() }
}

extern "C" {
  pub(self) fn tree_sitter_chip() -> Language;
}
