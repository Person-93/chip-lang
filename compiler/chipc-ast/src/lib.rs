pub use grammar::*;

#[path = "../grammar/bindings/rust/lib.rs"]
mod grammar;

#[cfg(test)]
mod tests {
  use crate::grammar;
  use insta::{assert_snapshot, glob};
  use std::{
    fmt::{self, Write},
    fs,
  };
  use tree_sitter::{Parser, Point, Tree, TreeCursor};

  #[test]
  fn snapshot() {
    let mut parser = Parser::new();
    parser
      .set_language(grammar::language())
      .expect("assign language to parser");

    glob!("cases/*.chip", |path| {
      parser.reset();
      let source = fs::read_to_string(path).expect("read snapshot test case");
      let tree = parser
        .parse(&source, None)
        .expect("parse snapshot test case");
      assert_snapshot!(print(&tree, &source));
    });
  }

  pub fn print(tree: &Tree, source: &str) -> String {
    let mut output = String::new();
    let mut cursor = tree.walk();
    output
      .print_node(&mut cursor, source, 0)
      .expect("debug format tree");
    output
  }

  impl<T: Write> NodePrinter for T {}

  trait NodePrinter: Write {
    fn print_node(
      &mut self,
      cursor: &mut TreeCursor,
      source: &str,
      indentation: u8,
    ) -> fmt::Result {
      let node = cursor.node();
      let kind = node.kind();
      let Point { row, column } = node.start_position();
      let Point {
        row: end_row,
        column: end_column,
      } = node.end_position();

      self.indent(indentation)?;
      write!(self, "{kind} @ {row}:{column}..{end_row}:{end_column}")?;
      if node.is_named() && node.child_count() == 0 {
        write!(self, "\t{}", node.utf8_text(source.as_bytes()).unwrap())?;
      }
      writeln!(self)?;

      if cursor.goto_first_child() {
        self.print_node(&mut cursor.clone(), source, indentation + 1)?;
        while cursor.goto_next_sibling() {
          self.print_node(&mut cursor.clone(), source, indentation + 1)?;
        }
      }

      Ok(())
    }

    fn indent(&mut self, indentation: u8) -> fmt::Result {
      for _ in 0..indentation {
        write!(self, "  ")?;
      }
      Ok(())
    }
  }
}
