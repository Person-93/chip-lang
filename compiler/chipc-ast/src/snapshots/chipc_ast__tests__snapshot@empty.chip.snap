---
source: compiler/chipc-ast/src/lib.rs
assertion_line: 27
expression: "print(&tree, &source)"
input_file: compiler/chipc-ast/src/cases/empty.chip
---
file @ 0:0..0:0
